src/Bedrock.Framework/Protocols/Protocol.cs

```diff
@@ -182,4 +184,18 @@ public interface IProtocolWriter<TMessage>
     {
         void WriteMessage(TMessage message, IBufferWriter<byte> output);
     }
+
+    public readonly struct ReadResult<TMessage>
+    {
+        public ReadResult(TMessage message, bool isCanceled, bool isCompleted)
```

Comment:
doesn't this causes `TMessage message` (that's a struct in the example) being copied into ReadResult?
should this be a ref?

Reply:
It does. I'm not sure the struct can be readonly and return a ref to the TMessage